Sources/SwiftKafka/KafkaPollingSystem.swift

```diff
@@ -0,0 +1,430 @@
+    var pollClosure: (() -> Void)? {
+        get {
+            self.stateMachine.withLockedValue { $0.pollClosure }
+        }
+        set {
+            self.stateMachine.withLockedValue { $0.pollClosure = newValue }
+        }
+    }
+
+    /// Initializes the ``KafkaPollingSystem``.
+    /// Private initializer. The ``KafkaPollingSystem`` is not supposed to be initialized directly.
+    /// It must rather be initialized using the ``KafkaPollingSystem.createSystemAndSequence`` function.
+    init() {
+        self.stateMachine = NIOLockedValueBox(StateMachine())
+    }
+
+    /// Runs the poll loop with the specified poll interval.
+    ///
+    /// - Parameter pollInterval: The desired time interval between two consecutive polls.
+    /// - Returns: An awaitable task representing the execution of the poll loop.
+    func run(pollInterval: Duration) async throws {
+        switch self.stateMachine.withLockedValue({ $0.run() }) {
+        case .alreadyClosed:
+            return
+        case .alreadyRunning:
+            fatalError("Poll loop must not be started more than once")
+        case .startLoop:
+            break
+        }
+
+        while true {
+            let action = self.nextPollLoopAction()
+
+            switch action {
+            case .pollAndSleep(let pollClosure):
+                // Poll Kafka for new acknowledgements and sleep for the given
+                // pollInterval to avoid hot looping.
+                pollClosure?()
+                do {
+                    try await Task.sleep(for: pollInterval)
+                } catch {
+                    let action = self.stateMachine.withLockedValue { $0.terminate() }
+                    self.handleTerminateAction(action)
+                    throw error
+                }
+            case .suspendPollLoop:
+                // The downstream consumer asked us to stop sending new messages.
+                // We therefore await until we are unsuspended again.
+                try await withTaskCancellationHandler {
+                    try await withCheckedThrowingContinuation { continuation in
+                        self.stateMachine.withLockedValue { $0.suspendLoop(continuation: continuation) }
+                    }
+                } onCancel: {
+                    let action = self.stateMachine.withLockedValue { $0.terminate(CancellationError()) }
+                    self.handleTerminateAction(action)
+                }
+            case .shutdownPollLoop:
+                // We have been asked to close down the poll loop.
+                let action = self.stateMachine.withLockedValue { $0.terminate() }
+                self.handleTerminateAction(action)
+            }
+        }
+    }
+
+    /// Yield new elements to the underlying `NIOAsyncSequenceProducer`.
+    func yield(_ element: Element) {
+        self.stateMachine.withLockedValue { stateMachine in
+            switch stateMachine.state {
```

Comment:
Why are we switching on the state here? We should switch on the state inside the state machine and return an action